Add Transfer-Encoding: chunked and avoid closing connections when possible

jekh opened the pull request:
This PR should address the issues discussed in #203. When an HTTP response is received that has no message-length indicator other than the server closing the connection, LittleProxy will now add a "Transfer-Encoding: chunked" header, which causes netty to automatically generate the proper chunked encoding. LittleProxy will also avoid closing connections to the client, even when all servers disconnect, so the client can re-use the connection for future requests.

ganskef commented:
While implementing test clients based on Netty I've noticed hanging connections (70s), after getting a chunked transmitted image via LittleProxy from a Netty based server. It depends on the server additionally (local Apache 15s). It's not happend with an offline (cached, not chunked) proxy response. It's not happens directly with no proxy, but chunked. It's not happens with a client based on java.net.URLConnection. That's my solution at the moment.

I had no ideas, but I could reproduce it, if you think this could help.

jekh commented:
@ganskef - do you mean you've experienced the issue you describe with this PR branch, or with the code currently in master?

ganskef commented:
I've noticed this in master, but I could try it again with your current 
modifications. It could be fixed now eventually.

Am 13.05.2015 um 20:30 schrieb Jason Hoetger:
> @ganskef <https://github.com/ganskef> - do you mean you've experienced
> the issue you describe with this PR branch, or with the code currently
> in master?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/adamfisk/LittleProxy/pull/204#issuecomment-101769185>.
>

jekh commented:
I added another commit that improves the detection of the final Transfer-Encoding type. I also removed detection for multipart/byteranges, which is now officially not supported as a means of detecting message length according to [RFC 7230 appendix A.2](https://tools.ietf.org/html/rfc7230#appendix-A.2), and in any case was hardly ever used.

jekh commented:
@ganskef -- I'd be very interested to see what your test scenarios are. Do you have some tests you can share, or example messages? After applying this PR and PR #198, if you are able to reproduce the issue you saw, it probably deserves a new LittleProxy issue.

ganskef commented:
Hello @jekh, please have a look at https://github.com/ganskef/LittleProxy-mitm/commit/429c58783cb1ce49c6e1caf2630a90f0055cbf99 .This is a copy of the URLConnection based test with a Netty based client. Please note, https is allways replaced with http here, but that's not relevant. It demonstrates the hanging connection after loading the content. Consider to modify the io.netty.handler.logging.LogLevel in the client and in the server. I hope, it's not a silly mistake, since I've tried it a while, with a lot of modifications.

ganskef commented:
Sorry, my Netty based test client implementation is wrong. Please don't investigate. 

Using the HttpSnoopClientInitializer from the Netty examples is not blocking. So, it was a silly mistake: I've missed to close the ChannelHandlerContext in the Handler on receiving LastHttpContent. I will fix that.

jekh commented:
Since this PR has been out for a while and has been reviewed without issues, and has solid unit test coverage, I'm going to merge this in.

ganskef commented:
Hi Jason, since I've merged the upstream, I noticed an issue: ganskef/LittleProxy-mitm#7.

Today, I've merged upstream again (#209), but the issue still arrives. Could you have a closer look at this please?

ganskef commented:
@jekh please double check 0f07bb34d784e2c2a9fc066f9f527dc33eb2de6b . This fixes ganskef/LittleProxy-mitm#7.
